### ISSUE TITLE
TDS Core Spinner Component does not prevent updating of fields while spinning

### ISSUE THREAD

**Issue by akdambenieks** [telus/tds-core#907]
## Description

TDS Core `Spinner` component:
While spinner is spinning it is still possible to keyboard tab into input fields to update values and buttons to resubmit events.

## Reproduction Steps

Enter input into a field and submit the form.  The spinner component wrapping the inputs receives the spinning attribute while a call to a rather slow API is made.  While the spinner is still 'spinning' it is possible to tab/shift+tab into the input fields to change their values or to tab onto the submit button and resubmit the form.  The spinner does however prevent mouse clicking into the wrapped inputs/buttons.

## Reproduction Percentage (if applicable)

Reproducible at least 9/9 times.  Others on my team as well as from DRB have no problem reproducing the issue.

## Meta

- TDS version: , Spinner version: 2.1.1
- Willing to develop solution: No
- Has workaround: Is possible
- High impact: Potentially yes, especially so for visually impaired users or for prolonged spinning events where repeated submission of forms can hang up the application.

## Screenshots

CODE USED TO IMPLEMENT THE SPINNER COMPONENT:
```jsx
    <FlexGrid limitWidth={false} gutter={false}>
        <Spinner inline spinning={isLoading || isLoadingAppointments} a11yLabel="Currently searching, please wait for results to be returned" tip="Working..">
          <FlexGrid.Row>
            <FlexGrid.Col>
              <Box between={4}>
                <Box between={3}>
                  {validationStatus === 'ERROR' &&
                    <Notification variant="error">
                      Please search and select an address. Orders cannot be created without a valid service address.
                    </Notification>
                  }
                  <Input
                    id="customerAddress"
                    value={address}
                    onChange={this._updateInput}
                    label="Street address"
                    aria-label="Customer's street address"
                    helper={AddressFormatHelper}
                  />
                  <Input
                    id="customerAddressInit"
                    value={addressUnit}
                    onChange={this._updateAddressUnit}
                    label="Unit number"
                    aria-label="Customer's unit number if applicable"
                    hint="if applicable"
                  />
                </Box>
                <Box>
                  <Button onClick={this._sendAutocomplete}>Search<A11yContent>for matching service address. Once results are returned please select one of the available addresses below</A11yContent></Button>
                </Box>
                <span aria-live="polite">
                  {addressStatus === 'NOMATCH' ?
                      'No results found for this address' : ''
                  }
                </span>
                {autocompleteList.length > 0 &&
                  <Box vertical={1}>
                    <Box tag="fieldset" between={2}>
                      <legend>
                        <Text bold size="medium">
                          Choose an address to use for this order
                        </Text>
                      </legend>
                      {autocompleteList.map(data => (
                        <ListItem
                          key={data.locationId}
                          data={data}
                          addressChoice={this.state.addressChoice}
                          onSelectCreator={this._updateSelection}
                        />
                      ))}
                    </Box>
                  </Box>
                }
              </Box>
            </FlexGrid.Col>
          </FlexGrid.Row>
        </Spinner>
      </FlexGrid>
```
Screenshot 1: Entering value into the input field and hitting submit button:
<img width="1440" alt="screen shot 2019-01-28 at 9 08 15 am" src="https://user-images.githubusercontent.com/22565360/51855487-3fcf1500-22e2-11e9-83a9-3563021c8db6.png">

Screenshot 2: Spinner now has `spinning` attribute:
<img width="1440" alt="screen shot 2019-01-28 at 9 08 43 am" src="https://user-images.githubusercontent.com/22565360/51855561-6ab96900-22e2-11e9-8740-f276f9a80814.png">

Screenshot 3: By tabbing/shift+tabbing it is possible to enter the input fields and update the values:
<img width="1440" alt="screen shot 2019-01-28 at 9 09 49 am" src="https://user-images.githubusercontent.com/22565360/51855622-8886ce00-22e2-11e9-838f-f886afd35a7b.png">

**Comment by simpleimpulse** [comment]
TDS core team to discuss with @owestinTelus  about accessibility concerns around disabling fields.

**Comment by mejsullivan** [comment]
Hello!

Who do I connect with regarding this issue? I think the priority might need altering.

Thank you,
Beth

**Comment by simpleimpulse** [comment]
Beth, I'd be happy to discuss more about the issue and how we can help resolve it. 
Please reach out on Slack, or email at anirudh.swaminathan@telus.com

**Comment by mejsullivan** [comment]
Thank you!

Get Outlook for Android<https://aka.ms/ghei36>




On Tue, Mar 12, 2019 at 10:58 PM -0700, "Ani" <notifications@github.com<mailto:notifications@github.com>> wrote:


Beth, I'd be happy to discuss more about the issue and how we can help resolve it.
Please reach out on Slack, or email at anirudh.swaminathan@telus.com<mailto:anirudh.swaminathan@telus.com>

**Comment by theetrain** [comment]
The core team will investigate Spinner, its usage, and aim to document best practices as well as provide necessary patches that address accessibility.

**Comment by theetrain** [comment]
This is being tracked on Jira and will be investigated in April: https://telusdigital.atlassian.net/browse/TDS-1246

**Comment by donnavitan** [comment]
h2. WIP notes
* Waiting for design review (visual updates to spinner when applied to button, content block, and full page)
* Updates to implementation, focus management, and usage criteria

h2. Implementation
* Prevent interactions with the content behind the spinner
* Include focus state on the spinner (includes overlay background, spinner, and label)
* Include spinner labels for action and result
* Spinner labels will be visually hidden when used on a button
* Spinner label will be displayed when used on a content block or full page


h2. Focus management 
* Object triggers spinner (user action)
* Spinner should have focus state, receive focus and announce spinner label (action)
* Wait for dynamic content to be done (no more interaction, data is fetched), announce spinner label (result)
* Return focus to:
** Origin (spawner object)
** Another target in close proximity, if it makes sense. For example, search results appearing after a search form.


h2. Usage criteria
* Use only for processes that take between 1 to 4 seconds to load (https://uxmovement.com/navigation/progress-bars-vs-spinners-when-to-use-which/)
* Spinner label should accurately explain the state of the requested action, “Loading”, “Submitting”, “Processing”. Be short and concise.
* Recommend not to use for full page loading

**Comment by donnavitan** [comment]
Latest updates:

### Visual design

- Background overlay
- White (#FFFFFF) at 94% opacity
- Expand 100% height and width of viewport
- Appears behind the spinner and caption
- Applies to both small and large

### Small spinner
- Apply to button 
- Visually hidden caption (a11y) 
- Spinner symbol
- - 24px height and width
- - 3px stroke
- - Available variants are primary (accessible green) and secondary (TELUS purple)
- Horizontally and vertically centered 


### Large spinner
- Apply to content block and full page 
- Spinner symbol
- - 48px height and width
- - 3px stroke 
- - Primary (accessible green) 
- Display caption 
- - Text is medium (16px) and centered 
- - Caption may expand up to 480px maximum width  (readability)
- - 48px left/right padding 
- Spinner and caption are horizontally and vertically centered

Specifications: https://telus.invisionapp.com/d/main#/console/13491097/359547132/preview

### Implementation
- Prevent interactions with the content behind the spinner
- Include focus state on the spinner (includes overlay background, spinner, and label)
- Include spinner labels for action and result
- Spinner labels will be visually hidden when used on a button
- Spinner label will be displayed when used on a content block or full page


### Focus management
- Object triggers spinner (user action)
- Spinner should have focus state, receive focus and announce spinner label (action)
- Wait for dynamic content to be done (no more interaction, data is fetched), announce spinner label (result)
- Return focus to:
- - Origin (spawner object)
- - Another target in close proximity, if it makes sense. For example, search results appearing after a search form.


### Usage criteria (Documentation)
- Label should be short and concise. Spinner label should accurately explain the state of the requested action, “Loading”, “Submitting”, “Processing”.
- Use only for processes that take between 1 to 4 seconds to load
- Use the small spinner when applied on a button
- Use the large spinner when affecting an entire web page or content block, whose content has not finished loading, such as paginated content
- Avoid using a spinner whenever possible. Make content available to the user as soon as possible.
- Recommend not to use for full page loading

**Comment by donnavitan** [comment]
As per discussion with @apurvray and Ani, we may update the small spinner to be applicable to smaller interactive elements beyond the button element. It can be applied to a download link that is making a back-end call. It can be used for toggle to indicate loading state.

Reviewing design with CX, will confirm once approved.

**Comment by donnavitan** [comment]
Discussed with @gkychow and it makes sense to him. Will reach out to DPA for consensus. https://telus.invisionapp.com/share/4ARREY09XVM

**Comment by pkandathil** [comment]
Was asked by @mejsullivan to inform the team that this fix is required for Sales Tool Portal (DRB Project) as well. Thank you for helping fix this issue.

**Comment by theetrain** [comment]
The Spinner component has been updated and now includes accessibility considerations within its documentation.

It covers use cases such as:

- overlaying buttons, links, or other interactive elements
- how to set up your project to announce loaded content
- covering purple or green buttons with higher contrast
- supplying a label to the Spinner using brief, relevant copy in any language

Learn more:

- [@tds/core-spinner@2.2.0](https://github.com/telus/tds-core/releases/tag/%40tds%2Fcore-spinner%402.2.0)
- [Spinner documentation](https://tds.telus.com/components/index.html#spinner)
- DSM Library version: _TDS_core v.2.1.10
- [DSM Release notes](https://telus.invisionapp.com/dsm/telus/tds-core/releases)

**Comment by mejsullivan** [comment]
@varunj90 should this not be re-opened?
@owestinTelus @dewing9

**Comment by varunj90** [comment]
The scope of this ticket is completed @mejsullivan. 
TDS core team will be reviewing 3 items from the accessibility matrix soon: https://telusdigital.atlassian.net/browse/TDS-1476

**Comment by mejsullivan** [comment]
Can you change the title of this issue to reflect what was fixed? When I did my validation the issue remained of people being able to get to the fields underneath the "loading..." ?

**Comment by varunj90** [comment]
Reopening issue: https://telusdigital.atlassian.net/browse/TDS-1567 in TDS backlog

**Comment by marcod1419** [comment]
This has been addressed in Spinner 3.1.0! https://github.com/telus/tds-core/releases/tag/%40tds%2Fcore-spinner%403.1.0
